Fix elastic stress used in principal_stress postprocessor

gassmoeller opened the pull request:
Similar to #4375, but for the principal_stress postprocessor. 

@rfildes: Could you check if this change also fixes the bug you observed? 

@bobmyhill or @anne-glerum: Could one of you review this PR?

bangerth commented on source/postprocess/visualization/principal_stress.cc:
```diff
@@ -115,36 +115,37 @@ namespace aspect
-                stress[1][1] += in.composition[q][this->introspection().compositional_index_for_name("ve_stress_yy")];
-                stress[0][1] += in.composition[q][this->introspection().compositional_index_for_name("ve_stress_xy")];
+                stress[0][0] = in.composition[q][this->introspection().compositional_index_for_name("ve_stress_xx")];
+                stress[1][1] = in.composition[q][this->introspection().compositional_index_for_name("ve_stress_yy")];
+                stress[0][1] = in.composition[q][this->introspection().compositional_index_for_name("ve_stress_xy")];
 
                 if (dim == 3)
                   {
-                    stress[2][2] += in.composition[q][this->introspection().compositional_index_for_name("ve_stress_zz")];
-                    stress[0][2] += in.composition[q][this->introspection().compositional_index_for_name("ve_stress_xz")];
-                    stress[1][2] += in.composition[q][this->introspection().compositional_index_for_name("ve_stress_yz")];
+                    stress[2][2] = in.composition[q][this->introspection().compositional_index_for_name("ve_stress_zz")];
+                    stress[0][2] = in.composition[q][this->introspection().compositional_index_for_name("ve_stress_xz")];
+                    stress[1][2] = in.composition[q][this->introspection().compositional_index_for_name("ve_stress_yz")];
                   }
               }
+            else
+              {
+                const SymmetricTensor<2,dim> deviatoric_strain_rate
+                  = (this->get_material_model().is_compressible()
+                     ?
+                     in.strain_rate[q] - 1./3. * trace(in.strain_rate[q]) * unit_symmetric_tensor<dim>()
+                     :
+                     in.strain_rate[q]);
+
+                // Compressive stress is positive in geoscience applications
+                stress = -2. * out.viscosities[q] * deviatoric_strain_rate;
+              }
```
Would you mind switching the if and else branches here? Most readers will approach this from the non-elasticity case in mind, and it would be nice to see the non-elastic case first.

rfildes replied:
I think also having the commented note about sign convention (the need to convert from the convention used inside of the code to that used in geology and in post processing here) moved above both the viscous and viscoelastic cases would make it clearer what is being done

bangerth reviewed (commented):
The postprocessor is documented as follows:
```
                                                  "A visualization output object that outputs the "
                                                  "principal stress values and directions, i.e., the "
                                                  "eigenvalues and eigenvectors of the stress tensor. "
                                                  "The postprocessor can either operate on the full "
                                                  "stress tensor or only on the deviatoric stress tensor, "
                                                  "depending on what run-time parameters are set."
```
I think it would be useful to define what "principal stress" actually means here, so that we can reason what the postprocessor should actually do. As it stands, it's hard to tell whether what you do is correct or not because we haven't *defined* what this postprocessor should actually do.

rfildes commented on source/postprocess/visualization/principal_stress.cc:
```diff
@@ -115,36 +115,37 @@ namespace aspect
-                stress[0][0] += in.composition[q][this->introspection().compositional_index_for_name("ve_stress_xx")];
-                stress[1][1] += in.composition[q][this->introspection().compositional_index_for_name("ve_stress_yy")];
-                stress[0][1] += in.composition[q][this->introspection().compositional_index_for_name("ve_stress_xy")];
+                stress[0][0] = in.composition[q][this->introspection().compositional_index_for_name("ve_stress_xx")];
+                stress[1][1] = in.composition[q][this->introspection().compositional_index_for_name("ve_stress_yy")];
+                stress[0][1] = in.composition[q][this->introspection().compositional_index_for_name("ve_stress_xy")];
```
```suggestion
                stress[0][0] = - in.composition[q][this->introspection().compositional_index_for_name("ve_stress_xx")];
                stress[1][1] = - in.composition[q][this->introspection().compositional_index_for_name("ve_stress_yy")];
                stress[0][1] = - in.composition[q][this->introspection().compositional_index_for_name("ve_stress_xy")];
```

rfildes commented on source/postprocess/visualization/principal_stress.cc:
```diff
@@ -115,36 +115,37 @@ namespace aspect
-                    stress[2][2] += in.composition[q][this->introspection().compositional_index_for_name("ve_stress_zz")];
-                    stress[0][2] += in.composition[q][this->introspection().compositional_index_for_name("ve_stress_xz")];
-                    stress[1][2] += in.composition[q][this->introspection().compositional_index_for_name("ve_stress_yz")];
+                    stress[2][2] = in.composition[q][this->introspection().compositional_index_for_name("ve_stress_zz")];
+                    stress[0][2] = in.composition[q][this->introspection().compositional_index_for_name("ve_stress_xz")];
+                    stress[1][2] = in.composition[q][this->introspection().compositional_index_for_name("ve_stress_yz")];
```
```suggestion
                    stress[2][2] = - in.composition[q][this->introspection().compositional_index_for_name("ve_stress_zz")];
                    stress[0][2] = - in.composition[q][this->introspection().compositional_index_for_name("ve_stress_xz")];
                    stress[1][2] = - in.composition[q][this->introspection().compositional_index_for_name("ve_stress_yz")];
```

rfildes commented:
@gassmoeller I have tested this and found it does not fix the stress orientation issue in the hanging beam test ( tests and description on this thread: https://community.geodynamics.org/t/principal-stress-orientations-in-viscous-vs-visco-elastic-models/3161/5).  I made some comments to your code changes above where I think the remaining issue is.

gassmoeller commented:
All good points, thanks for the review. I have corrected the signs and added explanation to the documentation of the postprocessor. The problem with the sign is really that it is a convention, and we just need to be consistent to use the same convention.
@rfildes do you want to contribute your test cases yourself in a separate PR (this would give you more credit for it), or do you want me to incorporate them into this PR?

rfildes commented:
@gassmoeller You can incorporate those test cases with this PR. I think that makes sense so it is all together! (I also was having some technical problems on my end and don't want to hold up the merging)

gassmoeller commented:
Ok, I added your tests to this PR and made sure the results are similar to the ones you posted in the forum.

This fix should be ready for a final review and merge.

bangerth commented:
One of the testers is stuck, but since the others are all succeeding, I'll take the liberty to merge.
